go/teams/chain.go

```diff
@@ -744,62 +744,62 @@ func (t *TeamSigChainPlayer) addInnerLink(
 			}
+		} else {
+			if prevState == nil {
+				return fmt.Errorf("link type '%s' unexpected at beginning", payload.Body.Type)
+			}
 		}
 		return nil
 	}
-	hasGeneric := func(hasExpected bool, hasReal bool, attr string) error {
-		if hasExpected != hasReal {
+	enforceGeneric := func(name string, rule Tristate, hasReal bool) error {
+		switch rule {
+		case TristateDisallow:
 			if hasReal {
-				return fmt.Errorf("team section contains unexpected %s", attr)
+				return fmt.Errorf("sigchain link contains unexpected '%s'", name)
+			}
+		case TristateRequire:
+			if !hasReal {
+				return fmt.Errorf("sigchain link missing '%s'", name)
 			}
-			return fmt.Errorf("missing %s", attr)
+		case TristateOptional:
+		default:
+			return fmt.Errorf("unsupported tristate (fault): %v", rule)
 		}
 		return nil
 	}
-	hasName := func(has bool) error {
-		return hasGeneric(has, team.Name != nil, "name")
-	}
-	hasMembers := func(has bool) error {
-		return hasGeneric(has, team.Members != nil, "members")
-	}
-	hasParent := func(has bool) error {
-		return hasGeneric(has, team.Parent != nil, "parent")
-	}
-	hasSubteam := func(has bool) error {
-		return hasGeneric(has, team.Subteam != nil, "subteam")
-	}
-	hasPerTeamKey := func(has bool) error {
-		return hasGeneric(has, team.PerTeamKey != nil, "per-team-key")
-	}
-	hasAdmin := func(has bool) error {
-		return hasGeneric(has, team.Admin != nil, "admin")
-	}
-	hasInvites := func(has bool) error {
-		return hasGeneric(has, team.Invites != nil, "invite")
-	}
-	hasCompletedInvites := func(has bool) error {
-		return hasGeneric(has, len(team.CompletedInvites) != 0, "completed_invites")
-	}
-	hasSettings := func(has bool) error {
-		return hasGeneric(has, team.Settings != nil, "settings")
-	}
-	hasKBFSSettings := func(has bool) error {
-		return hasGeneric(has, team.KBFS != nil, "KBFS settings")
-	}
-	allowInflate := func(allow bool) error {
-		if isInflate && !allow {
-			return fmt.Errorf("inflating link type not supported: %v", payload.Body.Type)
-		}
-		return nil
+	enforce := func(rules LinkRules) error {
+		return libkb.PickFirstError(
+			enforceGeneric("name", rules.Name, team.Name != nil),
+			enforceGeneric("members", rules.Members, team.Members != nil),
+			enforceGeneric("parent", rules.Parent, team.Parent != nil),
+			enforceGeneric("subteam", rules.Subteam, team.Subteam != nil),
+			enforceGeneric("per-team-key", rules.PerTeamKey, team.PerTeamKey != nil),
+			enforceGeneric("admin", rules.Admin, team.Admin != nil),
+			enforceGeneric("invites", rules.Invites, team.Invites != nil),
+			enforceGeneric("completed-invites", rules.CompletedInvites, team.CompletedInvites != nil),
+			enforceGeneric("settings", rules.Settings, team.Settings != nil),
+			enforceGeneric("kbfs", rules.KBFS, team.KBFS != nil),
+			allowInImplicitTeam(rules.AllowInImplicitTeam),
+			allowInflate(rules.AllowInflate),
+			enforceFirstInChain(!rules.FirstInChain),
```

Comment:
Is the `!` intentional? considering the CI failure probably not, and it's not immediately clear to me if it is

Reply:
Oh there it is! Absolutely not intentional, thanks